iina/Logger.swift

```diff
@@ -0,0 +1,146 @@
+    return sessionDir
+  }()
+
+  private static var logFileHandle: FileHandle = {
+    let logFileURL = logDirectory.appendingPathComponent("iina.log")
+    FileManager.default.createFile(atPath: logFileURL.path, contents: nil, attributes: nil)
+    return try! FileHandle(forWritingTo: logFileURL)
+  }()
+
+  private static var dateFormatter: DateFormatter = {
+    let formatter = DateFormatter()
+    formatter.dateFormat = "dd HH:mm:ss.SSS"
+    return formatter
+  }()
+
+  static func getLogger(_ label: String) -> Logger? {
+    return enabled ? Logger(label: label) : nil
+  }
+
+  static func closeLogFile() {
+    guard Logger.enabled else { return }
+    logFileHandle.closeFile()
+  }
+
+  private static func log(_ message: String, label: String, level: LogLevel, useNSLog: Bool, appendNewlineAtTheEnd: Bool) {
+    let time = dateFormatter.string(from: Date())
+    if useNSLog {
+      NSLog("[%@][%@] %@", label, level.description, message)
+      return
+    }
+    let string = "\(time) [\(label)][\(level.description)] \(message)\(appendNewlineAtTheEnd ? "\n" : "")"
+    if let data = string.data(using: .utf8) {
+      logFileHandle.write(data)
+    } else {
+      NSLog("Cannot encode log string!")
+    }
+  }
+
+  static let general: Logger? = {
+    #if DEBUG
+      let logger = Logger(label: "iina")
+      if !Logger.enabled {
+        logger.useNSLog = true
+      }
+      return logger
+    #else
+      return Logger.getLogger("iina")
+    #endif
+  }()
+
+  private init(label: String, logLevel: LogLevel? = nil) {
+    self.label = label
+    if let level = logLevel {
+      self.level = level
+    } else {
+      self.level = .preferenced
+    }
+  }
+
+  func verbose(_ message: String, appendNewline: Bool = true) {
+    guard level <= .verbose else { return }
+    Logger.log(message, label: label, level: .verbose, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func debug(_ message: String, appendNewline: Bool = true) {
+    guard level <= .debug else { return }
+    Logger.log(message, label: label, level: .debug, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func warning(_ message: String, appendNewline: Bool = true) {
+    guard level <= .warning else { return }
+    Logger.log(message, label: label, level: .warning, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  func error(_ message: String, appendNewline: Bool = true) {
+    guard level <= .error else { return }
+    Logger.log(message, label: label, level: .error, useNSLog: useNSLog, appendNewlineAtTheEnd: appendNewline)
+  }
+
+  static func assert(_ expr: Bool, _ errorMessage: String, _ block: () -> Void = {}) {
+    if !expr {
+      general?.error(errorMessage)
+      Utility.showAlert("fatal_error", arguments: [errorMessage])
+      block()
+      exit(1)
+    }
+  }
+
+  static func fatal(_ message: String, _ block: () -> Void = {}) -> Never {
+    general?.error(message)
```

Comment:
Here's an interesting thought: what if we threw an unhandled Objective-C exception here, and let AppKit handle showing a backtrace for us as part of the default `uncaughtErrorProc`?

Reply:
Won't IINA crash in that case? IMHO crashing indicates uncatched/unhandled exceptions of which the reason is needed to be investigated can can be fixed at the end, but here we call `fatal` only when there's an invalid operation (caused by user) and thus stop the App early to prevent future undefined behavior.